### ISSUE TITLE
Preventing specific messages using Middleware when subscribed to a channel

### ISSUE THREAD

**Issue by astutesoftware** [SocketCluster/socketcluster#344]
**I have a unique situation for a chat room.**

Tom, Dick, Harry and Jane are in a chat room.
They are subscribed to a channel for ex. "Lobby"

Dick being dick is very rude and Jane blocks him.
Now she should only be able to see messages from Tom and Harry.
Similarly Dick will be able to read messages from Tom and Harry but will not see anything from Jane.

How can this be achieved using the Middleware?

The server at this point at time knows who has blocked whom and the message payload does contain who is the sender - but i want the message not to be relayed to all subscribers and depending on the origin of the message (who sent it) I want specific subscribers to not receive the message.

**Comment by happilymarrieddad** [comment]
@astutesoftware store an array of "blocked users", maybe comma deliminated id's or something, in each user's row in your DB. Then you can simple emit to all clients who's ID is not in that array of blocked users. If there are no "blocked users" then it will broadcast to everyone.

**Comment by astutesoftware** [comment]
Yes you are right. That is exactly what i have in mind. My query was now that does the middleware expose the socket id of the person emitting the message to a channel and the socket id of the person going to receive the message from a channel.

For ex. At the middle ware level can i know the following
1. Dick has sent a message to the channel LOBBY. (I get to know Dicks' socket id which i can lookup in memory and map it and figure out that it is indeed Dick).
2. Jane is the intended recipient for the current message. (This is where i do the lookup to see if Jane indeed wants the receive the message from Dick, I get Jane's socket id in this case).

**Comment by MegaGM** [comment]
@astutesoftware For chat usecase, instead of blocking publish operations on the server side, you could simply hide the desired messages on the client side.
1. request from the server and keep a list of ignored userIDs on the client side (or, even better, store it in JWT)
2. skip the step in your app logic, which adds a recieved message in Jane's UI. Or you can still add the messages to the chat UI, but wrap them with a "spoiler tag", with a title "Blocked Message", or so.

The complex logic in the middlewares, for finding out if a message should be blocked or not, will cost you much more CPU (and headache), than publishing the message and do with it whatever you want on the client side :wink:

**Comment by happilymarrieddad** [comment]
@MegaGM That's a really good idea! You can just throw it out if as the client if it's not for you.
